packages/storage/amplify_storage_s3_dart/lib/src/model/s3_item.dart

```diff
@@ -0,0 +1,97 @@
+// Copyright 2022 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import 'package:amplify_core/amplify_core.dart';
+import 'package:amplify_storage_s3_dart/amplify_storage_s3_dart.dart';
+import 'package:amplify_storage_s3_dart/src/sdk/s3.dart' as s3;
+import 'package:meta/meta.dart';
+
+/// {@template storage.amplify_storage_s3.storage_s3_item}
+/// An object in a S3 bucket.
+/// {@endtemplate}
+class S3Item extends StorageItem {
+  /// {@macro storage.amplify_storage_s3.storage_s3_item}
+  const S3Item({
+    required super.key,
+    super.size,
+    super.lastModified,
+    super.eTag,
+    this.metadata = const <String, String>{},
+    this.versionId,
+  });
+
+  /// Creates a [S3Item] from [s3.S3Object] provided by S3 Client.
+  ///
+  /// This named constructor should be used internally only.
+  @internal
+  factory S3Item.fromS3Object(
+    s3.S3Object object, {
+    required String prefixToDrop,
+  }) {
+    final key = object.key;
+
+    // In S3 plugin, key is required property presenting an object
+    if (key == null) {
+      throw S3Exception.unknownException();
```

Comment:
Can you add some context to this exception? If this appeared in a log, it would be hard to diagnose. Maybe dump `object`?

Reply:
Adding null check here because the `s3.S3Object` model has nullable `key`, at this spot, `S3Object` is returned from S3 service, `key` should not be `null` (if null service fails the request and won't reach to this point), so throws this exception avoid using `!`.

Reply:
Understood. What I mean is that seeing "unknown exception" is not much more helpful than a null unwrap exception. Maybe change it to something like:

```dart
throw S3Exception.unknownException('Missing key in object: $object');
```

so that at least there's a bit of context when diagnosing.